src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java

```diff
@@ -0,0 +1,197 @@
+import uk.gov.hmcts.reform.bulkscan.payment.processor.service.servicebus.model.PaymentMessage;
+
+import java.time.Instant;
+
+import static uk.gov.hmcts.reform.bulkscan.payment.processor.service.servicebus.handler.MessageProcessingResultType.POTENTIALLY_RECOVERABLE_FAILURE;
+import static uk.gov.hmcts.reform.bulkscan.payment.processor.service.servicebus.handler.MessageProcessingResultType.SUCCESS;
+import static uk.gov.hmcts.reform.bulkscan.payment.processor.service.servicebus.handler.MessageProcessingResultType.UNRECOVERABLE_FAILURE;
+
+
+@Service
+@Profile("!nosb") // do not register for the nosb (test) profile
+public class PaymentMessageProcessor {
+
+    private static final Logger log = LoggerFactory.getLogger(PaymentMessageProcessor.class);
+
+    private final PaymentMessageHandler paymentMessageHandler;
+    private final IMessageReceiver messageReceiver;
+    private final int maxDeliveryCount;
+
+    public PaymentMessageProcessor(PaymentMessageHandler paymentMessageHandler,
+                                   IMessageReceiver messageReceiver,
+                                   @Value("${azure.servicebus.payments.max-delivery-count}") int maxDeliveryCount) {
+        this.paymentMessageHandler = paymentMessageHandler;
+        this.messageReceiver = messageReceiver;
+        this.maxDeliveryCount = maxDeliveryCount;
+    }
+
+
+    /**
+     * Reads and processes next message from the queue.
+     *
+     * @return false if there was no message to process. Otherwise true.
+     */
+    public boolean processNextMessage() throws ServiceBusException, InterruptedException {
+        IMessage message = messageReceiver.receive();
+        if (message != null) {
+            MessageProcessingResult result = process(message);
+            tryFinaliseProcessedMessage(message, result);
+        }
+        return message != null;
+    }
+
+    private MessageProcessingResult process(IMessage message) {
+        log.info("Started processing message with ID {}", message.getMessageId());
+
+        PaymentMessage payment = null;
+
+        try {
+            payment = PaymentMessageParser.parse(message.getMessageBody());
+            logMessageParsed(message, payment);
+            paymentMessageHandler.handlePaymentMessage(payment);
+            log.info("Processed message with ID {}. Envelope ID: {}", message.getMessageId(), payment.envelopeId);
+            return new MessageProcessingResult(SUCCESS);
+        } catch (InvalidMessageException ex) {
+            log.error("Rejected message with ID {}, because it's invalid", message.getMessageId(), ex);
+            return new MessageProcessingResult(UNRECOVERABLE_FAILURE, ex);
+        } catch (PayHubClientException ex) {
+            logMessageProcessingError(message, payment, ex);
+            if (ex.getStatus() == HttpStatus.BAD_REQUEST.CONFLICT) {
+                return new MessageProcessingResult(UNRECOVERABLE_FAILURE, ex);
```

Comment:
Should it be success? 409 from PayHub means that this operation has already been completed.

Reply:
Then, if it's a 400/422, then it should be unrecoverable failure. For other cases, it should be potentially recoverable failure.

Reply:
it was successful before , now as it is 4xx it should be sth else to distinguish ?
we can change the enum names?

Reply:
Not sure I understand your question.

Reply:
I mean, may be enum names are confusing, 409 is not success Http wise,

Reply:
We might think about enum names, but then they should ideally be the same in the orchestrator and here.

Regarding 409, we told Payment team that their operations had to be idempotent. In response, they told us that PayHub would return 409 in case it has already processed given request. If we get 409, we can treat this as success (but log a warnings) and continue.